pkg/compactor/block_upload_test.go

```diff
@@ -1259,23 +1240,257 @@ func TestMultitenantCompactor_HandleBlockUpload_Complete(t *testing.T) {
+				bucketClient: injectedBkt,
+				cfgProvider:  cfgProvider,
+			}
+			userBkt := bucket.NewUserBucketClient(tenantID, injectedBkt, cfgProvider)
+
+			meta := metadata.Meta{}
+			marshalAndUploadJSON(t, bkt, uploadingMetaPath, meta)
+			v := validationFile{}
+			marshalAndUploadJSON(t, bkt, validationPath, v)
+
+			c.validateAndCompleteBlockUpload(log.NewNopLogger(), userBkt, ulid.MustParse(blockID), &meta, tc.validation)
+
+			tempUploadingMetaExists, err := bkt.Exists(context.Background(), uploadingMetaPath)
+			require.NoError(t, err)
+			require.Equal(t, tempUploadingMetaExists, tc.expectTempUploadingMeta)
+
+			metaExists, err := bkt.Exists(context.Background(), metaPath)
+			require.NoError(t, err)
+			require.Equal(t, metaExists, tc.expectMeta)
+
+			if !tc.expectValidationFile {
+				exists, err := bkt.Exists(context.Background(), validationPath)
+				require.NoError(t, err)
+				require.False(t, exists)
+				return
+			}
+
+			r, err := bkt.Get(context.Background(), validationPath)
+			require.NoError(t, err)
+			decoder := json.NewDecoder(r)
+			err = decoder.Decode(&v)
+			require.NoError(t, err)
+
+			if tc.expectErrorInValidationFile {
+				require.NotEmpty(t, v.Error)
+			} else {
+				require.Empty(t, v.Error)
+			}
+		})
+	}
+}
+
+func TestMultitenantCompactor_PeriodicValidationUpdater(t *testing.T) {
+	const tenantID = "test"
+	const blockID = "01G3FZ0JWJYJC0ZM6Y9778P6KD"
+	injectedError := fmt.Errorf("injected error")
+	validationPath := path.Join(tenantID, blockID, validationFilename)
+
+	heartbeatInterval := 50 * time.Millisecond
+
+	validationExists := func(t *testing.T, bkt objstore.Bucket) bool {
+		exists, err := bkt.Exists(context.Background(), validationPath)
+		require.NoError(t, err)
+		return exists
+	}
+
+	testCases := []struct {
+		name          string
+		errorInjector func(op bucket.Operation, name string) error
+		cancelContext bool
+		assertions    func(t *testing.T, ctx context.Context, bkt objstore.Bucket)
+	}{
+		{
+			name:          "updating validation file fails",
+			errorInjector: bucket.InjectErrorOn(bucket.OpUpload, validationPath, injectedError),
+			assertions: func(t *testing.T, ctx context.Context, bkt objstore.Bucket) {
+				<-ctx.Done()
+				require.True(t, errors.Is(context.Canceled, ctx.Err()))
+				require.False(t, validationExists(t, bkt))
+			},
+		},
+		{
+			name: "updating validation file succeeds",
+			assertions: func(t *testing.T, ctx context.Context, bkt objstore.Bucket) {
+				time.Sleep(heartbeatInterval)
+				test.Poll(t, heartbeatInterval*2, true, func() interface{} {
+					return validationExists(t, bkt)
+				})
+
+				v := validationFile{}
+				r, err := bkt.Get(context.Background(), validationPath)
+				require.NoError(t, err)
+				decoder := json.NewDecoder(r)
+				err = decoder.Decode(&v)
+				require.NoError(t, err)
+				require.NotEqual(t, 0, v.LastUpdate)
+				require.Empty(t, v.Error)
+			},
+		},
+		{
+			name: "context cancelled before update",
```

Comment:
Shall we set `cancelContext: true` for this test?

To make sure this test fails without `cancelContext: true`, shall we sleep for one heartbeat interval (only if `cancelContext == false`) before calling `assertions` function?

Reply:
> Shall we set cancelContext: true for this test?

Ugh, thanks!

> shall we sleep for one heartbeat interval (only if cancelContext == false) before calling assertions function

Sure, added.